### PR TITLE
[DPB] Allow BGP_NEIGHBOR YANG model with just IP as the key

### PR DESCRIPTION
**Pull request by venkatmahalingam**
#### Why I did it
[DPB] BGP_NEIGHBOR yang model causing DPB errors.

#### How I did it
Allowed the BGP_NEIGHBOR list with just IP as the key to allow the current BGP neighbor table in the community.

#### How to verify it

Built the SONIC YANG models.

### CODE REVIEW

**Review by robocoder99 — CHANGES_REQUESTED** [review]
Please see inline feedback.

File: src/sonic-yang-models/tests/files/sample_config_db.json
Comment (robocoder99) [inline src/sonic-yang-models/tests/files/sample_config_db.json:869]: Having two `BGP_NEIGHBOR` entries in this file will not work. You cannot have two of the same keys in most softwares, this will attract bugs. (I just tested this in redis and it does not work)

File: src/sonic-yang-models/tests/files/sample_config_db.json
Comment (venkatmahalingam) [reply src/sonic-yang-models/tests/files/sample_config_db.json:881]: will add few fields here, but we have the UT to cover multiple fields.

File: src/sonic-yang-models/tests/files/sample_config_db.json
Comment (venkatmahalingam) [reply src/sonic-yang-models/tests/files/sample_config_db.json:869]: will have 2 entries in BGP_NEIGHBOR table one with just IP and another one with VRF & IP to ensure both the tables parsed without any issues.

File: src/sonic-yang-models/yang-models/sonic-bgp-neighbor.yang
Comment (venkatmahalingam) [reply src/sonic-yang-models/yang-models/sonic-bgp-neighbor.yang:55]: I dont understand this command, what's hardcoded here?

File: src/sonic-yang-models/yang-models/sonic-bgp-neighbor.yang
Comment (venkatmahalingam) [reply src/sonic-yang-models/yang-models/sonic-bgp-neighbor.yang:52]: Yes, I'm aware of that, this should be done for all tables based on frr_mgmt_framework_config, will do it.

File: src/sonic-yang-models/yang-models/sonic-bgp-neighbor.yang
Comment (robocoder99) [reply src/sonic-yang-models/yang-models/sonic-bgp-neighbor.yang:55]: You use `uses bgpcmn:sonic-bgp-cmn;` to get the individual attributes for `BGP_NEIGHBOR_LIST` but you individually list out every attribute for `BGP_NEIGHBOR_TEMPLATE_LIST`

Was asking about why this was done differently for the two lists.

Comment (venkatmahalingam) [reply src/sonic-yang-models/yang-models/sonic-bgp-neighbor.yang:55]: We cant use 'sonic-bgp-cmn' as it is specific to new BGP_NEIGHBOR table that's handled by frrcfgd, I want to it be separate to avoid any conflict.

File: src/sonic-yang-models/tests/files/sample_config_db.json
Comment (venkatmahalingam) [reply src/sonic-yang-models/tests/files/sample_config_db.json:869]: I understand that we can't have two entries (one with just IP and another one with VRF+IP as keys) together, will check with @praveen-li  on how to accommodate two different configs in the sample_config_db.json if required.

**Comment by venkatmahalingam** [comment]
@robocoder99 I think, this change can be merged to unblock your DPB testing, we can sort out the pending items based on the infra (libyang) capability.

**Comment by praveen-li** [comment]
Btw, Which application is using bgp yang model today?

File: src/sonic-yang-models/tests/files/sample_config_db.json
Comment (praveen-li) [reply src/sonic-yang-models/tests/files/sample_config_db.json:869]: We do not need 2 BGP_NEIGHBOR tables in config?, As per yang models, it is same container, which allows both type of entries.

**Comment by robocoder99** [comment]
Only dynamic port breakout.

**Comment by praveen-li** [comment]
> Only dynamic port breakout.

Then I am not sure, why  these yang models are not written as per config_db entries?

**Comment by robocoder99** [comment]
> 
> 
> > Only dynamic port breakout.
> 
> Then I am not sure, why these yang models are not written as per config_db entries?

Not sure what you mean here, can you elaborate?

**Comment by robocoder99** [comment]
@praveen-li The two forms of Yang model are for two different bgp daemons is what my impression is. Where each daemon takes different configuration format. What I do know is that this yang model was originally written with a different configuration format in mind than what SONiC loads by default on master right now. I am working under the assumption that we need to support both formats depending on the `frr_mgmt_framework_config` value. 

@venkatmahalingam Its fine with me if this gets pushed as a hotfix to stabilize DPB on platforms not using the frr form of the DPB config, if its good with other maintainers. However, we do need a long-term solution for this yang model.

**Comment by venkatmahalingam** [comment]
@praveen-li @robocoder99 If you don't have any other comments, please approve this PR and merge the same.

File: src/sonic-yang-models/tests/files/sample_config_db.json
Comment (venkatmahalingam) [inline src/sonic-yang-models/tests/files/sample_config_db.json:872]: "BGP_NEIGHBOR": {
            "default|192.168.1.1": {
                "asn": "65200",
                "holdtime": "180",
                "keepalive": "60",
                "local_addr": "10.0.0.2",
                "name":"PEER2"
            }
        },

@praveen-li These fields are present in the grouping, looks like libyang is not able to parse it.

 Read JSON Section: SAMPLE_CONFIG_DB_JSON
sonic_yang(3):All Keys are not parsed in BGP_NEIGHBOR
dict_keys(['default|192.168.1.1'])
sonic_yang(3):exceptionList:['Value not found for neighbor in default|192.168.1.1', "'asn'"]
sonic_yang(3):Data Loading Failed:All Keys are not parsed in BGP_NEIGHBOR
dict_keys(['default|192.168.1.1'])

Comment (praveen-li) [reply src/sonic-yang-models/tests/files/sample_config_db.json:872]: @venkatmahalingam:  Yeah, [current translation doesn't allow all the variation of leaf, only choice and leaf is allowed.] if you are using grouping, please add support at:
https://github.com/Azure/sonic-buildimage/blob/5acf2348ba88c69bad3300adabd58b270e5bc01f/src/sonic-yang-mgmt/sonic_yang_ext.py#L209 

Also, about the other BGP_NEIGHBOR table
 -- where the relevant config go, as I understand BGP_NEIGHBOR table in config_db is different.
 -- which daemon processes it, does it make use of current YANG models. 
 -- please point to the code, which is using this BGP_NEIGHBOR yang model.

Comment (venkatmahalingam) [reply src/sonic-yang-models/tests/files/sample_config_db.json:872]: New generic BGP config models are being processed by the below frrcfgd daemon in BGP container.
https://github.com/Azure/sonic-buildimage/blob/master/src/sonic-frr-mgmt-framework/frrcfgd/frrcfgd.py

File: src/sonic-yang-models/yang-models/sonic-bgp-neighbor.yang
Comment (praveen-li) [inline src/sonic-yang-models/yang-models/sonic-bgp-neighbor.yang:77]: Should we make sure that local address and Peer address belongs to same address family?

File: src/sonic-yang-models/yang-models/sonic-bgp-neighbor.yang
Comment (praveen-li) [inline src/sonic-yang-models/yang-models/sonic-bgp-neighbor.yang:53]: Can we add example config in sample config_db.json

File: src/sonic-yang-models/yang-models/sonic-bgp-neighbor.yang
Comment (praveen-li) [inline src/sonic-yang-models/yang-models/sonic-bgp-neighbor.yang:87]: Ideally BGP neighbor config must be on below page:
https://github.com/Azure/SONiC/wiki/Configuration

@alexrallen : would you like to add BGP neighbor config section on this page.
FYI: @lguohan

File: src/sonic-yang-models/yang-models/sonic-bgp-neighbor.yang
Comment (venkatmahalingam) [reply src/sonic-yang-models/yang-models/sonic-bgp-neighbor.yang:53]: I did in my original commit but Alex mentioned both BGP_NEIGHBOR based on frrcfgd and bgpcfgd cant co-exist and hence I removed it.

File: src/sonic-yang-models/yang-models/sonic-bgp-neighbor.yang
Comment (venkatmahalingam) [reply src/sonic-yang-models/yang-models/sonic-bgp-neighbor.yang:77]: I dont see any easy way to check this, do you have any easy check in mind?

File: src/sonic-yang-models/yang-models/sonic-bgp-neighbor.yang
Comment (alexrallen) [reply src/sonic-yang-models/yang-models/sonic-bgp-neighbor.yang:53]: My specific complaint was the organization of the json file having two keys. One would have overwritten the other and it would not have tested well. 

Given this is a unit test I think it is acceptable to have both forms of the config to validate that both forms are correctly parsed / recongnized by the model.

File: src/sonic-yang-models/yang-models/sonic-bgp-neighbor.yang
Comment (alexrallen) [reply src/sonic-yang-models/yang-models/sonic-bgp-neighbor.yang:87]: I can bring it up to date.

**Comment by venkatmahalingam** [comment]
@praveen-li Can you look at my replies for your comments, if you are okay, can you approve this PR?

**Comment by venkatmahalingam** [comment]
@alexrallen If you don't have any more comments, please approve this PR.

**Review by alexrallen — APPROVED** [review]
This looks fine to me. 

With the current state of this PR it **will** allow Yang to accept the bgpcfgd form of CONFIG_DB neighbor entries which will fix the root cause of #7959 so given that I would support merging it. 

However....
1. The frrcfgd form is still broken and will be until our Yang implementation is updated to accept the `grouping` keyword per @praveen-li comment. This is an outstanding bug. 
2. We still need to intelligently differentiate between what form of BGP config the user is using via the DEVICE_METADATA table. This can be a future enhancement. 

I will refactor issue #7959 to scope only for the bgpcfgd issue so it may be closed following merging this PR in recognition that the default BGP neighbor template in SONiC should now work and allow DPB to function. 

However, I will also raise a new bug to ensure that the issues with the `grouping` keyword are tracked.

File: src/sonic-yang-models/tests/files/sample_config_db.json
Comment (praveen-li) [reply src/sonic-yang-models/tests/files/sample_config_db.json:872]: @venkatmahalingam: can you please point to the function or relevant line in frrcfgd.py, where it is doing config validation against yang. Also from where it reads the config i.e. source of config?.

Comment (venkatmahalingam) [reply src/sonic-yang-models/tests/files/sample_config_db.json:872]: @praveen-li There is no config validation against YANG in frrcfgd, frrcfgd simply listens for config-DB events and handles them, all configs validations are expected to handle in SONiC YANGs before pushing to config-DB.

**Review by praveen-li — APPROVED** [review]
I still do not fully understand, why there are 2 tables\Lists in Yang model, when other table is not used in any YANG validation. But I am approving it based on manual testing done by NVDIA Devs\Tests.

**Comment by venkatmahalingam** [comment]
> I still do not fully understand, why there are 2 tables\Lists in Yang model, when other table is not used in any YANG validation. But I am approving it based on manual testing done by NVDIA Devs\Tests.

Let's discuss this in the next week's YANG meeting.

**Comment by venkatmahalingam** [comment]
@lguohan Please help to merge this PR.

**Comment by venkatmahalingam** [comment]
@anshuv-mfst @lguohan Please merge this PR.

**Comment by venkatmahalingam** [comment]
@anshuv-mfst Please help to merge this PR.

**Comment by dgsudharsan** [comment]
@anshuv-mfst Is there anything that we are still waiting on this PR? If not can you please help merge this?

**Comment by praveen-li** [comment]
Hi Sudharsan

Are these PR approved and we are waiting just for Merge, or we are waiting from review as well?

From: Sudharsan Dhamal Gopalarathnam ***@***.***>
Date: Tuesday, August 10, 2021 at 3:19 PM
To: Azure/sonic-buildimage ***@***.***>
Cc: Praveen Chaudhary ***@***.***>, Mention ***@***.***>
Subject: Re: [Azure/sonic-buildimage] [DPB] Allow BGP_NEIGHBOR YANG model with just IP as the key (#7992)

@anshuv-mfst<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fanshuv-mfst&data=04%7C01%7Cpchaudhary%40linkedin.com%7C5f014ec1379c4fae911208d95c4cf137%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637642307826325686%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=zKFKxmsj8RCPwYnX73kQCNCCzXLvWpMAjUGLTZoe94Y%3D&reserved=0> Is there anything that we are still waiting on this PR? If not can you please help merge this?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fsonic-buildimage%2Fpull%2F7992%23issuecomment-896349871&data=04%7C01%7Cpchaudhary%40linkedin.com%7C5f014ec1379c4fae911208d95c4cf137%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637642307826335642%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=GL6MeSrjntNvfm4cCphdcP2bpvXMQ%2Fq1%2F9lB5XM9wzI%3D&reserved=0>, or unsubscribe<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAJHZHPMT7CNKUGAOOWPPPW3T4GQXNANCNFSM47NFPE6A&data=04%7C01%7Cpchaudhary%40linkedin.com%7C5f014ec1379c4fae911208d95c4cf137%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637642307826335642%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=trXUcAdyy79dUc8sndtrNPkJyLc4zjoTNnr1hL9%2FH5s%3D&reserved=0>.
Triage notifications on the go with GitHub Mobile for iOS<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fapps.apple.com%2Fapp%2Fapple-store%2Fid1477376905%3Fct%3Dnotification-email%26mt%3D8%26pt%3D524675&data=04%7C01%7Cpchaudhary%40linkedin.com%7C5f014ec1379c4fae911208d95c4cf137%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637642307826335642%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=ysB0vOtKvDwddnW4f6l3cCgYf99OXxx%2F%2FYsUhEyO4SM%3D&reserved=0> or Android<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fplay.google.com%2Fstore%2Fapps%2Fdetails%3Fid%3Dcom.github.android%26utm_campaign%3Dnotification-email&data=04%7C01%7Cpchaudhary%40linkedin.com%7C5f014ec1379c4fae911208d95c4cf137%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637642307826345597%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=FPOttAGBuy1YBLwAYrEGFZAI5MkPlir7Bbr2ZVvEKV0%3D&reserved=0>.

**Comment by dgsudharsan** [comment]
@praveen-li I am not sure what is the criteria for merge. That's why I asked @anshuv-mfst. If we need some more approvals who should be reviewing it?

**Comment by anshuv-mfst** [comment]
Hi @shi-su - could you please review, approve and help with merge. Thanks!

**Review by shi-su — APPROVED** [review]

**Pull request merged**
